src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -632,46 +632,43 @@ private void CreateGrainInstance(string grainTypeName, ActivationData data, stri
 
             GrainTypeData grainTypeData = GrainTypeManager[grainClassName];
 
+            //Get the grain's type
             Type grainType = grainTypeData.Type;
 
-            //Create a new instance of the given grain type
-            var grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
-            grain.Data = data;
-
+            //Gets the type for the grain's state
             Type stateObjectType = grainTypeData.StateObjectType;
-            object state = null;
-            if (stateObjectType != null)
-            {
-                state = Activator.CreateInstance(stateObjectType);
-            }
             
             lock (data)
             {
-                data.SetGrainInstance(grain);
-                var statefulGrain = grain as IStatefulGrain;
-                if (statefulGrain != null)
+                Grain grain;
+
+                //Create a new instance of a stateless grain
+                if (stateObjectType == null)
                 {
-                    if (state != null)
-                    {
-                        SetupStorageProvider(data);
-                        statefulGrain.GrainState.State = state;
-                        statefulGrain.SetStorage(new GrainStateStorageBridge(data.GrainTypeName, statefulGrain,
-                            data.StorageProvider));
-                    }
+                    //Create a new instance of the given grain type
+                    grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
```

Comment:
If as you said the bigger lock is not needed, I'd reduce it to what it was before, and keep instantiation out of it